DAOS-16834 test: tags.py config option

phender opened the pull request:
Add config command line option to tags.py to list test yaml data.

Test-tag: always_passes
Skip-unit-tests: true
Skip-fault-injection-test: true

### Steps for the author:

* [ ] Commit message follows the [guidelines](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Appropriate [Features or Test-tag](https://daosio.atlassian.net/wiki/spaces/DC/pages/10984259629/Test+Tags) pragmas were used.
* [ ] At least two positive code reviews including at least one code owner from each category referenced in the PR.
* [ ] Testing is complete. If necessary, forced-landing label added and a reason added in a comment.

#### After all prior steps are complete:
* [ ] Gatekeeper requested (daos-gatekeeper added as a reviewer).

daltonbohning committed:
DAOS-15630 test: tags.py dump tests associated with tags

Add support in tags.py to dump tests associated with tags and/or paths.
Add TagSet to handle negative matching. E.g. foo,-bar

Test-tag: always_passes
Skip-unit-tests: true
Skip-fault-injection-test: true

Signed-off-by: Dalton Bohning <dalton.bohning@hpe.com>

daltonbohning committed:
Merge remote-tracking branch 'origin/master' into dbohning/daos-15630

daltonbohning committed:
optimization

Test-tag: always_passes
Skip-unit-tests: true
Skip-fault-injection-test: true

Signed-off-by: Dalton Bohning <dalton.bohning@hpe.com>

daltonbohning committed:
handle incompatible args

Test-tag: always_passes
Skip-unit-tests: true
Skip-fault-injection-test: true

Signed-off-by: Dalton Bohning <dalton.bohning@hpe.com>

daltonbohning committed:
lint fix

Test-tag: always_passes
Skip-unit-tests: true
Skip-fault-injection-test: true

Signed-off-by: Dalton Bohning <dalton.bohning@hpe.com>

daltonbohning committed:
optimize

Test-tag: always_passes
Skip-unit-tests: true
Skip-fault-injection-test: true

Signed-off-by: Dalton Bohning <dalton.bohning@hpe.com>

phender committed:
DAOS-16834 test: tags.py data option

Add data command line option to tags.py to list test yaml data.

Test-tag: always_passes
Skip-unit-tests: true
Skip-fault-injection-test: true

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender committed:
Fix lint.

Test-tag: always_passes
Skip-unit-tests: true
Skip-fault-injection-test: true

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

daltonbohning commented on src/tests/ftest/tags.py:
```diff
@@ -650,84 +726,206 @@ def print_verbose(*args):
+        print(f"Generated {csv_file} with {len(output)} records")
+
+    else:
+        width = []
+        for line in output:
+            for index, column in enumerate(line):
+                size = len(str(column))
+                if index == len(width):
+                    width.append(size)
+                elif size > width[index]:
+                    width[index] = size
```
I'd recommend simplifying since we know how many there will be upfront
```suggestion
        widths = [0] * len(output[0])
        for line in output:
            for index, column in enumerate(line):
                widths[index] = max(widths[index], len(str(column)))
```

phender replied:
Updated, thanks.

daltonbohning commented on src/tests/ftest/tags.py:
```diff
@@ -650,84 +726,206 @@ def print_verbose(*args):
+                    width[index] = size
+        format_line = "  ".join(list(map(lambda x: f"{{:{x}}}", width)))
+        for line in output:
+            print(format_line.format(*list(map(lambda x: str(x), line))))
```
pylint is complaining because lambda isn't needed
```suggestion
            print(format_line.format(*list(map(str, line))))
```

phender replied:
Done.

daltonbohning commented on src/tests/ftest/tags.py:
```diff
@@ -650,84 +726,206 @@ def print_verbose(*args):
+
+    # Parser for the "data" command and its optional arguments
+    data_parser = subparsers.add_parser(
+        'data',
```
I think something like "config" or "parse-config", "dump-config" etc. is more clear than "data", which is generic.

phender replied:
Updated.

daltonbohning commented on src/tests/ftest/tags.py:
```diff
@@ -650,84 +726,206 @@ def print_verbose(*args):
+    unit_parser.add_argument(
+        "-v", "--verbose",
+        action='store_true',
+        help="print verbose output for some commands")
```
We should only include the verbose argument for commands that use it
```suggestion
        help="print verbose output")
```

phender replied:
Done.

daltonbohning commented on src/tests/ftest/tags.py:
```diff
@@ -650,84 +726,206 @@ def print_verbose(*args):
+    Returns:
+        int: 0 = success
+             1 = error
+             2 = code problem
```
I don't think we should return rc 2. If there is a code issue we should just raise an exception.

phender replied:
Add exception in place of rc=2.

daltonbohning commented on src/tests/ftest/tags.py:
```diff
@@ -267,6 +287,62 @@ def __parse_avocado_tags(text):
         return set(','.join(tag_strings).split(','))
 
 
+class TestYamlData():
```
Maybe `TestConfig`? "data" is very generic. Everything is data :)

phender replied:
Sounds good to me.

phender committed:
Replace import from util.

Test-tag: always_passes
Skip-unit-tests: true
Skip-fault-injection-test: true

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender committed:
Apply feedback.

Test-tag: always_passes
Skip-unit-tests: true
Skip-fault-injection-test: true

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender committed:
Mised one.

Test-tag: always_passes
Skip-unit-tests: true
Skip-fault-injection-test: true

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender committed:
Fix test_frequency.

Test-tag: always_passes
Skip-unit-tests: true
Skip-fault-injection-test: true

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender commented on src/tests/ftest/tags.py:
```diff
@@ -267,6 +287,62 @@ def __parse_avocado_tags(text):
+            value = self.__get_test_method_value("timeouts", int)
+            if value == []:
+                # Handle special case for test-specific string timeout values
+                value = self.__get_test_method_value("timeouts", str)
```
Addressed, thanks.

phender commented on src/tests/ftest/tags.py:
```diff
@@ -650,84 +726,206 @@ def print_verbose(*args):
+        return "daily"
+    elif "weekly_regression" in tags:
+        return "weekly"
+    return "manual"
```
Fixed, thanks.

daltonbohning commented on src/tests/ftest/tags.py:
```diff
@@ -650,90 +747,210 @@ def print_verbose(*args):
+        help="print verbose output")
+
     args = parser.parse_args()
     args.paths = list(map(os.path.realpath, args.paths))
```
This doesn't work now with the subparsers that don't have paths. E.g.
```
./src/tests/ftest/tags.py unit
Traceback (most recent call last):
  File "/home/dbohning/daos-stack/daos/./src/tests/ftest/tags.py", line 959, in <module>
    sys.exit(main())
  File "/home/dbohning/daos-stack/daos/./src/tests/ftest/tags.py", line 933, in main
    args.paths = list(map(os.path.realpath, args.paths))
AttributeError: 'Namespace' object has no attribute 'paths'
```

phender replied:
Ah, right. I meant to make that conditional.

daltonbohning commented on src/tests/ftest/tags.py:
```diff
@@ -650,90 +747,210 @@ def print_verbose(*args):
+        return "weekly"
+    if "manual" in tags:
+        return "manual"
+    return ""
```
Maybe "unknown" or raise an exception? Not a big deal though

phender committed:
Fix --paths handling.

Test-tag: always_passes
Skip-unit-tests: true
Skip-fault-injection-test: true

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

daltonbohning commented:
You'll want to merge in master now that mine is merged

phender committed:
Merge branch 'master' into hendersp/tags-data

Test-tag: test_always_passes
Skip-unit-tests: true
Skip-fault-injection-test: true

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

daltonbohning commented on src/tests/ftest/tags.py:
```diff
@@ -267,6 +285,85 @@ def __parse_avocado_tags(text):
+        Returns:
+            list: values found for the key
+        """
+        search = [[self.__data, ["root"]]]
```
What is "root" for?

daltonbohning commented on src/tests/ftest/tags.py:
```diff
@@ -650,90 +747,211 @@ def print_verbose(*args):
         test_tag_set()
         test_tags_util(args.verbose)
         rc = 0
+    else:
+        raise ValueError(f"Unhandled command option: {args.command}")
```
I get this
```
./tags.py 
Traceback (most recent call last):
  File "/home/dbohning/daos-stack/daos/src/tests/ftest/./tags.py", line 960, in <module>
    sys.exit(main())
  File "/home/dbohning/daos-stack/daos/src/tests/ftest/./tags.py", line 954, in main
    raise ValueError(f"Unhandled command option: {args.command}")
ValueError: Unhandled command option: None
```

phender replied:
Fixed by making subparser 'command' required.

daltonbohning commented on src/tests/ftest/tags.py:
```diff
@@ -650,90 +747,211 @@ def print_verbose(*args):
+             1 = error
+    """
+    parser = ArgumentParser(prog='tags')
+    subparsers = parser.add_subparsers(
+        title='options for the tags command',
+        dest='command')
```
We need `required=True` since it is required

phender replied:
It does not appear so:
```
$ ./tags.py
usage: tags.py [-h] [-v] [--paths PATHS [PATHS ...]] [--tags TAGS [TAGS ...]] {lint,list,dump,unit}
tags.py: error: the following arguments are required: command

phender replied:
Yes, thanks.

shimizukko commented on src/tests/ftest/tags.py:
```diff
@@ -650,90 +747,211 @@ def print_verbose(*args):
+    Args:
+        paths (list, optional): paths to files from which to list via their tags. Defaults to all
+            ftest python files.
+        list (set, optional): list of sets of tags used to filter displayed tests. Defaults to no
```
```suggestion
        tags (set, optional): list of sets of tags used to filter displayed tests. Defaults to no
```
Typo?

phender replied:
Yes, thanks.

shimizukko commented on src/tests/ftest/tags.py:
```diff
@@ -650,90 +747,211 @@ def print_verbose(*args):
+            ftest python files.
+        list (set, optional): list of sets of tags used to filter displayed tests. Defaults to no
+            filtering.
+        list (set, optional): list of sets of test yaml data keys to display. Defaults to None,
```
```suggestion
        keys (set, optional): list of sets of test yaml data keys to display. Defaults to None,
```

phender replied:
Thanks.

phender committed:
Support paths in config search.

Test-tag: test_always_passes
Skip-unit-tests: true
Skip-fault-injection-test: true

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

daltonbohning commented on src/tests/ftest/tags.py:
```diff
@@ -267,6 +285,102 @@ def __parse_avocado_tags(text):
+        Returns:
+            list: values found for the key
+        """
+        self.__display(f"__filter_value: key_path={key_path}, key_types={key_types}")
```
I think it's better to not leave this debug here. Especially since it requires a helper method + new argument in init. This is only ever needed if there is a bug. And to guard against bugs its better to have unit tests than debug printouts
